presto-main/src/main/java/io/prestosql/event/QueryMonitor.java

```diff
@@ -339,6 +371,58 @@ private static QueryIOMetadata getQueryIOMetadata(QueryInfo queryInfo)
         return new QueryIOMetadata(inputs.build(), output);
     }
 
+    private static Multimap<QualifiedObjectName, OperatorStats> extractTableScanOperatorStats(QueryInfo queryInfo)
+    {
+        // Note: A plan may scan a table multiple times.
+        ImmutableMultimap.Builder<QualifiedObjectName, OperatorStats> tableScanOperatorStats = ImmutableMultimap.builder();
+        getAllStages(queryInfo.getOutputStage())
+                .forEach(stageInfo -> extractTableScanOperatorStats(stageInfo, tableScanOperatorStats));
+        return tableScanOperatorStats.build();
+    }
+
+    private static void extractTableScanOperatorStats(StageInfo stageInfo, ImmutableMultimap.Builder<QualifiedObjectName, OperatorStats> tableScanOperatorStats)
+    {
+        if (stageInfo.getPlan() == null) {
+            return;
+        }
+
+        // Note: a plan node may be mapped to multiple operators.  For this code it doesn't matter
+        // since only the table scan operator has physical stats, but be careful for any other use.
+        Map<PlanNodeId, Collection<OperatorStats>> allOperatorStats = Multimaps.index(stageInfo.getStageStats().getOperatorSummaries(), OperatorStats::getPlanNodeId).asMap();
+
+        // Find the operator stats for each table scanned.  Sometimes a table scan is merged with
+        // other operators, and in that case, use the stats of the nearest parent node that with stats.
+        stageInfo.getPlan().getRoot().accept(
+                new PlanVisitor<Void, Collection<OperatorStats>>()
+                {
+                    @Override
+                    protected Void visitPlan(PlanNode node, Collection<OperatorStats> parentStats)
+                    {
+                        for (PlanNode child : node.getSources()) {
+                            child.accept(this, allOperatorStats.getOrDefault(node.getId(), parentStats));
+                        }
+                        return null;
+                    }
+
+                    @Override
+                    public Void visitTableScan(TableScanNode node, Collection<OperatorStats> parentStats)
```

Comment:
should we also add the same logic for `IndexSourceScanNode` like in `InputExtractor`?